ownCloud/Media Uploads/PhotoKit Extensions/PHAsset+Upload.swift

```diff
@@ -456,41 +466,56 @@ extension PHAsset {
                 fileName = "\(self.ocStyleUploadFileName(with: sourceURL.deletingPathExtension().lastPathComponent.imageFileNameSuffix)).\(sourceURL.pathExtension)"
 			}
 
-			// Synchronously import media file into the OCCore and schedule upload
-			let importSemaphore = DispatchSemaphore(value: 0)
-
-			uploadProgress = sourceURL.upload(with: core,
-											  at: rootItem,
-											  alternativeName: fileName,
-											  modificationDate: self.creationDate,
-											  importByCopy: copySource,
-											  cellularSwitchIdentifier: cellularSwitchIdentifier,
-											  placeholderHandler: { (item, error) in
-												if !copySource && error != nil {
-													// Delete the temporary asset file in case of critical error
-													removeSourceFile()
-												}
-												if error != nil {
-													Log.error(tagged: ["MEDIA_UPLOAD"], "Sync engine import failed for asset ID \(self.localIdentifier)")
-												} else {
-													Log.debug(tagged: ["MEDIA_UPLOAD"], "Finished uploading asset ID \(self.localIdentifier)")
-												}
-												importResult = (item, error)
-												importSemaphore.signal()
-												uploadCompleteHandler?()
-
-			}, completionHandler: { (_, _) in
+			if let fpSession = fpSession {
+				fpSession.importThroughFileProvider(url: sourceURL, as: fileName, to: rootItem) { error in
+					importResult = (nil, error)
```

Comment:
@felix-schwarz yes... true but at least if there is an error it won't get lost completely and is going to be logged at least

Reply:
When following the call hierarchy, the returned `(OCItem?, Error?)` result ends up as having a meaningful impact in `MediaUploadOperation.swift:109`:

```swift
// Perform asset import
if let itemLocalId = self.importAsset(asset: asset,
									  with: core,
									  at: item,
									  uploadCompletion: {
	// Import successful
	self.removeUploadJob(with: path)
})?.localID as OCLocalID? {

	// Update media upload storage object
	core.bookmark.modifyMediaUploadStorage { (storage) in
		storage.update(localItemID: itemLocalId, assetId: self.assetId, targetPath: path)
		return storage
	}
}
```

If `(nil, nil)` is returned as `importAsset` result here, the `assetId`/`itemLocalId` pair will not be saved … couldn't that lead to photo asset uploads not being tracked and/or being uploaded more than once?

Reply:
@felix-schwarz indeed.. not sure what would be the best way to deal with that, since `OCFileProviderServiceSession.importThroughFileProvider()` is sort of fire-and-forget API which doesn't guarantee that the item is going to be uploaded and corresponding `OCItem` is going to exist, right?

Reply:
@mneuwert `OCFileProviderServiceSession.importThroughFileProvider()` opens an XPC connection to the File Provider, calls a special import method in the FP and only returns via the completion handler after the (local) import of the file either succeeded or has failed for any reason.

What did not get returned until now, however, was the newly created item itself, as the `OCItem` could quickly be outdated. Since background uploads only need the `OCLocalID` and that ID doesn't change, I extended the completion handler with a second argument: the `OCLocalID`.

I have now:
- adapted all code using the upload through File Provider to the change
- extended the use of the semaphore to the part using `OCFileProviderServiceSession.importThroughFileProvider()`, so a correct result is returned
- made sure that (now) returned local ID is returned as part of the result

That should address the issues. From my side, the only missing part now is testing.

Please also have a look at the changes I made, in case I overlooked something important.